### PR TITLE
Add sklearn regressors, optimize hyperparameter search space and enhance cross-validation

### PR DESCRIPTION
**Pull request by lishen**
Contributions in this PR:

- Add sklearn regressors.
- Optimize hyperparameter search space for classifiers/regressors so that the learners give good performance in most cases.
- Add K-fold, leave-one-out, shuffle-and-split cross-validation with shuffle option and stratification for classification.
- Add a lag selector as preprocessing for time series forecasting problems.
- Refactoring to improve code modularity and readability.
- Fix obsolete tests so that they run under current version of hpsklearn.
- Misc. bug fixes and enhancements.

### CODE REVIEW

**Comment by lishen** [comment]
@jaberg Please offer some insights as to why TestSpace did not work. 

@jaberg @bjkomer I can switch from skdata to sklearn.datasets for the test_demo. I think it will improve the reliability of the test code. Please let me know if you want me to do so.

**Comment by lishen** [comment]
I know I have made a lot of changes to the code. Please let me know if there is anything I can do to make it clearer. I can add some additional tests for the regression and other changes. I can also do some rebasing to clean up some commits. Let me know what you think. @jaberg @bjkomer

**Comment by jaberg** [comment]
This is exciting! I will look at it some time this week, please ping me if
I forget.

**Comment by lishen** [comment]
Hi @jaberg, what's the status of code review? Any feedback? Thanks!

Li

**Comment by jaberg** [comment]
Hi Li!

I will try to make time for this next week, but I'm not actively
maintaining or using this code these days. Maybe @bjkomer will have some
thoughts about it too? What do you think Brent?

**Comment by bjkomer** [comment]
Lots of improvements here, great job! I'll look through these changes and try them out more this weekend but I'm not as familiar with the code as I used to be.

I like your idea of switching from skdata to sklearn.datasets for the tests, as sometimes problems with importing it arise, and less dependencies is always nice.

**Comment by lishen** [comment]
@jaberg @bjkomer 
Some updates here:
- I added back the test_demo.py and changed the Iris dataset from skdata to sklearn.
- I changed the display of the loss function in the demo from graphical to textual output. I tried to use the original code, which is supposed to show a plot of the losses vs. iteration but was not successful. Changing the output to textual is much easier and should be universally available on all platforms. It also makes the code less dependent on other packages.
- I added a regression demo using the sklearn Boston dataset so you can be sure the regressors are working happily with hyperopt.
- I tried rebasing some of the commits but wasn't successful because there are parallel histories in the repo. So I just leave the commits as is. The only thing I tried to rebase is the time series lag selector because I changed the design. Anyway, the code works fine. I just tried to make the commits look more concise.

**Comment by bjkomer** [comment]
Looks great!! All the tests are working fine on my machine, and all of my old code still works. Nothing is jumping out at me that I think needs to be changed.

**Comment by lishen** [comment]
Just added AdaBoost and GradientBoosting regressors. Fixed a few minor bugs. 

I had to tune down some parameters for the test_preproc in test_stories.py because it is taking too long to finish. If I use trial_timeout, the preprocessing may not finish properly, which will affect follow-up training and testing. 

All tests passed. @bjkomer @jaberg

**Comment by lishen** [comment]
Added a new test for time series forecast. This new test serves two purposes:

1. Illustrate the use of lag selectors and exogenous data.
2. Demonstrate a way to format time series data for use with sklearn + hyperopt. 

I hope these additional tests and modifications will strengthen the reliability of the code in this PR. If nothing else needs to be changed, I hope you can accept this PR since I believe many people can benefit from the new additions. @bjkomer @jaberg

**Comment by bjkomer** [comment]
Excellent work! You're right, a lot of people can benefit from this so I'll get it merged now. If anything comes up that needs to be changed that can always be dealt with in another PR.
